Cookie compression and splitting for JWT

jochen-kressin opened the pull request:
### Description
This PR will port the cookie compression and splitting behaviour as implemented already for OIDC and SAML to also be used for the JWT authentication type.
This is done to support token payloads that are too large to store in a single cookie.

### Category
Enhancement

### Why these changes are required?
When the user authenticates with a JWT, the token is stored in a cookie. However, if the JWT is too large to store in a single cookie, the authentication will fail. To remedy this, we implemented the same approach as we've previously done for SAML and OpenId: compression and cookie splitting.



### What is the old behavior before changes and new behavior after changes?
The old behaviour was that large JWTs would cause the authentication to fail. Now, we support larger tokens by splitting it up in multiple cookies, the number of which can be defined in the configuration.

For end users that are already authenticated, existing cookies should continue to work.


### Issues Resolved
https://github.com/opensearch-project/security-dashboards-plugin/issues/1638

### Testing
[Please provide details of testing done: unit testing, integration testing and manual testing]
Manual testing:
I've tested with large JWTs, and I've tested that having an existing cookie still works after the splitting was implemented.
I did start testing multi auth, but then quickly noticed that JWT isn't supported in multi auth

Unit tests:
I had to edit the existing JWT unit tests a bit. I did not change any logic, but only extended certain "mocks" with properties required by the new implementation.

Integration tests:
Since there already were Selenium tests for JWT, I stuck with that and added a test with a JWT that is too large for one cookie.

### Additional notes
Even though we now support large tokens in the cookies, the user may still run into a problem with a "Too long http header exception".
```
{"statusCode":401,"error":"Unauthorized","message":"[too_long_http_header_exception] HTTP header is larger than 8192 bytes."}
```
This can however be configured in opensearch.yml, e.g. with `http.max_header_size: 16kb`
I have not investigated if this change has any negative side effects.


### Check List
- [x] New functionality includes testing
- [ ] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

jochen-kressin commented on server/auth/types/jwt/jwt_auth.ts:
```diff
@@ -25,10 +25,21 @@ import {
+  setExtraAuthStorage,
+} from '../../../session/cookie_splitter';
+
+export const JWT_DEFAULT_EXTRA_STORAGE_OPTIONS: ExtraAuthStorageOptions = {
```
I added this const here because of the JWT config schema being defined as `maybe`.
So this means that if no config for JWT is present in `opensearch_dashboards.yml`, there won't be any JWT values available in the config passed to the plugin, despite default values being used.

So that means that a customer may have JWT defined as authentication type, but unless they've changed any JWT config values, the plugin won't have access to the (default) config.
For SAML and OIDC we didn't want to make the cookie splitting configurable, so I went the same path here. And to make sure we don't run into problems when setting up the cookies, I used this constant as a fallback for the default values. You can see the usage on line 92 in this file.

I think we could just remove the `maybe` from the JWT config schema, but I didn't want to change too much without knowing the reason for it being there in the first place.

jochen-kressin commented:
Another thing I found while testing this was that the logout doesn't work for JWT.
I believe it is just the logout path that is wrong in the JWT routes file.
Didn't want to update it in the scope of this issue, in case there are any edge cases I don't know about.
But feel free to ping me if you want me to change it in this PR.

scrawfor99 commented:
Hi @jochen-kressin, for the logout issue--do you think you could open an issue or separate PR to fix it? Thank you!

cwperks commented:
Thank you for this PR @jochen-kressin! While I was reviewing this [issue](https://github.com/opensearch-project/security-dashboards-plugin/issues/1621) I noticed it happened because the token was too large. This PR should fix that.

jochen-kressin commented:
> Hi @jochen-kressin, for the logout issue--do you think you could open an issue or separate PR to fix it? Thank you!

@cwperks Will do! While testing I also noticed that the SAML logout throws an error. Will check again if I configured something wrong, otherwise I'll add an issue for that as well.

jochen-kressin commented:
While adding the logout issue, I realized I had forgotten something related to this issue.
Just as for SAML and OIDC, we should clear the split cookies when logging out. I've added that now.

I added https://github.com/opensearch-project/security-dashboards-plugin/issues/1710 for the logout issue. I considered fixing the issue within this PR, but then thought it best to keep it separate so that we can get this PR merged.

The SAML logout issue I mentioned: I believe that was a misconfiguration on my part, so I'm not adding an issue for that.
The issue I had was that the IdP would redirect back to the logout route after I was already logged out, resulting in a 401 error. But that redirect is configurable with the IdP, so I just updated it there.
